src/lib.rs

```diff
@@ -17,6 +18,228 @@ mod session;
 
 use crate::panic::PanicOrDefault;
 
+// For C callbacks, we need to offer a `void *userdata` parameter, so the
+// application can associate callbacks with particular pieces of state. We
+// allow setting a userdata pointer on a per-session basis, but the rustls
+// session objects don't offer a way to store a `c_void` attached to a session.
+// So we use thread-locals. Before calling out to rustls code that may call
+// a callback, we set USERDATA for the current thread to the userdata pointer
+// for the current session. Before returning to the C caller, we restore
+// USERDATA to its previous value. Because a C callback may call back into
+// Rust code, we model these thread locals as a stack, so we can always
+// restore the previous version.
+thread_local! {
+    pub static USERDATA: RefCell<Vec<*mut c_void>> = RefCell::new(Vec::new());
+}
+
+/// UserdataGuard pops an entry off the USERDATA stack, restoring the
+/// thread-local state to its value previous to the creation of the UserdataGuard.
+/// Invariants: As long as a UserdataGuard is live:
+//// - The stack of userdata items for this thread must have at least one item.
+///  - The top item on that stack must be the one this guard was built with.
+///  - The `data` field must not be None.
+/// If any of these invariants fails, try_drop will return an error.
+pub struct UserdataGuard {
+    // Keep a copy of the data we expect to be popping off the stack. This allows
+    // us to check for consistency, and also serves to make this type !Send:
+    // https://doc.rust-lang.org/nightly/std/primitive.pointer.html#impl-Send-1
+    data: Option<*mut c_void>,
+}
+
+impl UserdataGuard {
+    fn new(u: *mut c_void) -> Self {
+        UserdataGuard { data: Some(u) }
+    }
+
+    /// Even though we have a Drop impl on this guard, when possible it's
+    /// best to call try_drop explicitly. That way any failures of internal
+    /// variants can be signaled to the user immediately by returning
+    /// rustls_result::Panic.
+    fn try_drop(mut self) -> Result<(), UserdataError> {
+        self.try_pop()
+    }
+
+    fn try_pop(&mut self) -> Result<(), UserdataError> {
+        let expected_data = match self.data {
+            Some(d) => d,
+            None => return Err(UserdataError::AlreadyPopped),
+        };
+        eprintln!("popping");
```

Comment:
I don't think libraries should clutter client's stderr with debug output by default. Rich logging options are more API surface than `crustls` needs but could this be restricted to debug builds?

Reply:
Oops, leftover debugging output. Thanks for catching.